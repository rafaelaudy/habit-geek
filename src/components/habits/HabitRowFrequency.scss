@import "../Variables.scss";

.habit-row-frequency {
  width: 380px;
  display: flex;
  justify-content: space-between;

  &__cell {
    width: 55px;
    text-align: center;

    &--today {
      border: $border-width solid $muted;
      border-width: 0 $border-width;
    }

    .toggle {
      margin: 5px 0 5px 6px;
    }
  }
}

@media (max-width: 576px) {
  .habit-row-frequency {
    order: 1;

    &__cell {
      font-size: 1rem;
      padding-top: 5px;
      width: 46px;

      &--today {
        border: $border-width solid $muted;
        border-radius: 0.25rem;
        padding-top: 4px;
      }
    }
  }
}
