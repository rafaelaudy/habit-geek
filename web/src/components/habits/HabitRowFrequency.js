import React from "react";
import {
  getTodayIndex,
  getCurrentWeek,
  isToday
} from "@habit-geek/shared/utils/dateUtils";

import HabitCheckbox from "./HabitCheckbox";
import "./HabitRowFrequency.scss";

const HabitRowFrequency = ({
  week,
  name,
  checked,
  habitSucceded,
  habitFailed,
  isReadOnly,
  toggleDayHabit
}) => {
  const todayIndex = getTodayIndex();
  const currentWeek = getCurrentWeek();
  const isCurrentWeek = week === currentWeek;
  const succededClass = habitSucceded ? "table-success" : "";
  const failedClass = habitFailed ? "table-danger" : "";

  return (
    <div className={`habit-row-frequency ${succededClass} ${failedClass}`}>
      {[...Array(7).keys()].map(checkIndex => {
        const todayClass = isToday(week, checkIndex)
          ? "habit-row-frequency__cell--today"
          : "";
        const isTodayOrBefore = todayIndex < checkIndex;
        const isDisabled = isCurrentWeek ? isTodayOrBefore : false;

        return (
          <div
            className={`habit-row-frequency__cell ${todayClass}`}
            key={`check-${name}-${checkIndex}`}
          >
            <HabitCheckbox
              clickHandler={() => toggleDayHabit(week, name, checkIndex)}
              isChecked={checked[checkIndex]}
              isDisabled={isDisabled}
              isReadOnly={isReadOnly}
            ></HabitCheckbox>
          </div>
        );
      })}
    </div>
  );
};

export default HabitRowFrequency;
