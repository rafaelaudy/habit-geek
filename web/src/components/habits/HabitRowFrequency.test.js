import React from "react";
import { shallow, mount } from "enzyme";
import HabitRowFrequency from "./HabitRowFrequency";
import { isToday } from "@habit-geek/shared/utils/dateUtils";

jest.mock("@habit-geek/shared/utils/dateUtils.js", () => ({
  getTodayIndex: jest.fn().mockReturnValue(1),
  getCurrentWeek: jest.fn().mockReturnValue("y1w1"),
  isToday: jest.fn().mockReturnValue(false)
}));

const checked = [true, true, false, false, true, false, false];

describe("HabitRowFrequency component", () => {
  it("Renders static elements", () => {
    const habitRowFrequency = shallow(
      <HabitRowFrequency
        week="week"
        name="name"
        checked={checked}
        habitSucceded={false}
        habitFailed={false}
        isReadOnly={true}
        toggleDayHabit={() => {}}
      />
    );
    expect(habitRowFrequency).toMatchSnapshot();
  });

  it("marks habits performing", () => {
    const habitRowFrequency = shallow(
      <HabitRowFrequency checked habitSucceded={true} habitFailed={true} />
    );
    expect(habitRowFrequency.find(".table-danger").length).toBe(1);
    expect(habitRowFrequency.find(".table-success").length).toBe(1);
  });

  it("shows border around current date", () => {
    isToday.mockReturnValueOnce(true);
    const habitRowFrequency = shallow(<HabitRowFrequency checked={checked} />);
    expect(
      habitRowFrequency.find(".habit-row-frequency__cell--today").length
    ).toBe(1);
  });

  it("disables checkboxes for dates in the future", () => {
    const habitRowFrequency = shallow(
      <HabitRowFrequency week="y1w1" checked={checked} />
    );
    expect(habitRowFrequency.find("[isDisabled=true]").length).toBe(5);
  });

  it("do not disable checkboxes in previous weeks", () => {
    const habitRowFrequency = shallow(
      <HabitRowFrequency week="y1w0" checked={checked} />
    );
    expect(habitRowFrequency.find("[isDisabled=true]").length).toBe(0);
  });

  it("toggles habits", () => {
    const toggleDayHabitMock = jest.fn();
    const habitRowFrequency = mount(
      <HabitRowFrequency
        checked={checked}
        toggleDayHabit={toggleDayHabitMock}
      />
    );
    habitRowFrequency
      .find("input")
      .at(0)
      .simulate("change");
    expect(toggleDayHabitMock).toHaveBeenCalledTimes(1);
  });
});
