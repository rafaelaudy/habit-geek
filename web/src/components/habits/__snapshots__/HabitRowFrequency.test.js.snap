// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HabitRowFrequency component Renders static elements 1`] = `
<div
  className="habit-row-frequency  "
>
  <div
    className="habit-row-frequency__cell "
    key="check-name-0"
  >
    <HabitCheckbox
      clickHandler={[Function]}
      isChecked={true}
      isDisabled={false}
      isReadOnly={true}
    />
  </div>
  <div
    className="habit-row-frequency__cell "
    key="check-name-1"
  >
    <HabitCheckbox
      clickHandler={[Function]}
      isChecked={true}
      isDisabled={false}
      isReadOnly={true}
    />
  </div>
  <div
    className="habit-row-frequency__cell "
    key="check-name-2"
  >
    <HabitCheckbox
      clickHandler={[Function]}
      isChecked={false}
      isDisabled={false}
      isReadOnly={true}
    />
  </div>
  <div
    className="habit-row-frequency__cell "
    key="check-name-3"
  >
    <HabitCheckbox
      clickHandler={[Function]}
      isChecked={false}
      isDisabled={false}
      isReadOnly={true}
    />
  </div>
  <div
    className="habit-row-frequency__cell "
    key="check-name-4"
  >
    <HabitCheckbox
      clickHandler={[Function]}
      isChecked={true}
      isDisabled={false}
      isReadOnly={true}
    />
  </div>
  <div
    className="habit-row-frequency__cell "
    key="check-name-5"
  >
    <HabitCheckbox
      clickHandler={[Function]}
      isChecked={false}
      isDisabled={false}
      isReadOnly={true}
    />
  </div>
  <div
    className="habit-row-frequency__cell "
    key="check-name-6"
  >
    <HabitCheckbox
      clickHandler={[Function]}
      isChecked={false}
      isDisabled={false}
      isReadOnly={true}
    />
  </div>
</div>
`;
